### PR TITLE
Local execution skeleton

### PR DESCRIPTION
**Pull request by aitorrf**
- Audio configuration
- Models and libraries loaded
- LocalServiceImpl is a temporal solution

### CODE REVIEW

File: app/src/main/java/com/voysis/android/MainActivity.kt [hunk 1]
```diff
@@ -19,21 +20,16 @@ import com.voysis.events.FinishedReason
...
+import com.voysis.recorder.AudioRecordParams
 import com.voysis.sevice.DataConfig
-import kotlinx.android.synthetic.main.activity_main.cancel
-import kotlinx.android.synthetic.main.activity_main.eventText
-import kotlinx.android.synthetic.main.activity_main.responseText
-import kotlinx.android.synthetic.main.activity_main.send
-import kotlinx.android.synthetic.main.activity_main.start
-import kotlinx.android.synthetic.main.activity_main.stop
-import kotlinx.android.synthetic.main.activity_main.textInput
+import kotlinx.android.synthetic.main.activity_main.*
```
Comment (oranoceallaigh) [inline app/src/main/java/com/voysis/android/MainActivity.kt]: Still a star import hanging around.

File: app/src/main/java/com/voysis/android/MainActivity.kt [hunk 2]
```diff
@@ -19,21 +20,16 @@ import com.voysis.events.FinishedReason
...
 import java.net.URL
 import java.util.concurrent.Executors
 
 class MainActivity : AppCompatActivity(), Callback {
 
     private val url = "INSERT_URL"
-    private val config = DataConfig(isVadEnabled = true, url = URL(url), refreshToken = "INSERT_TOKEN", userId = "")
+    private val config = DataConfig(isVadEnabled = true, url = URL(url), refreshToken = "INSERT_TOKEN", userId = "",audioRecordParams = AudioRecordParams(sampleRate = 16000), offline =  true)// TODO delete offline true
```
Comment (oranoceallaigh) [inline app/src/main/java/com/voysis/android/MainActivity.kt]: Did you mean to leave this todo comment?

File: app/src/main/java/com/voysis/android/MainActivity.kt [hunk 3]
```diff
@@ -71,7 +67,9 @@ class MainActivity : AppCompatActivity(), Callback {
     }
 
     private fun init() {
-        service = ServiceProvider().make(applicationContext, config)
+        if(!::service.isInitialized)
+            service = ServiceProvider().make(applicationContext, config)
+        main.visibility=View.VISIBLE//FIRST VERSION OF LOADING, it hides the view until service is ready
```
Comment (oranoceallaigh) [inline app/src/main/java/com/voysis/android/MainActivity.kt]: Code style: spaces around `=` and no trailing-line comments (put the comment on its own line if it's needed)

File: app/src/main/java/com/voysis/android/MainActivity.kt (hunk 2 above)
Comment (ttroy50) [reply app/src/main/java/com/voysis/android/MainActivity.kt]: spacing of some of these look a bit off. Tere is a missing space before `audioRecordParams` and an extra one after `offline =`

File: library/src/main/java/com/voysis/api/Config.kt [hunk 4]
```diff
@@ -41,4 +40,8 @@ interface Config {
      * @return audioRecordParams
      */
     val audioRecordParams: AudioRecordParams?
+    /*
+     *
```
Comment (ttroy50) [inline library/src/main/java/com/voysis/api/Config.kt]: Missing a description of what this value does. Is it described somewhere else?

File: library/src/main/java/com/voysis/Utils.kt [hunk 5]
```diff
@@ -100,10 +100,17 @@ fun generateAudioRecordParams(context: Context, config: Config): AudioRecordPara
     val readBufferSize = generateReadBufferSize(config)
     val recordBufferSize = config.audioRecordParams?.recordBufferSize
             ?: AudioRecorderImpl.DEFAULT_RECORD_BUFFER_SIZE
+    val rate = generateRate(context, config)
+    return AudioRecordParams(rate, readBufferSize, recordBufferSize)
+}
 
+fun generateRate(context: Context, config: Config): Int? {
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/Utils.kt]: Please add kotlic-docs for new functions you're adding.

File: library/src/main/java/com/voysis/api/ServiceProvider.kt [hunk 6]
```diff
@@ -31,8 +37,10 @@ class ServiceProvider {
              okClient: OkHttpClient? = null,
              audioRecorder: AudioRecorder = AudioRecorderImpl(generateAudioRecordParams(context, config))): Service {
         val converter = Converter(getHeaders(context), Gson())
-        val client = createClient(config, generateOkHttpClient(okClient), converter)
-        return ServiceImpl(client, audioRecorder, converter, config.userId, TokenManager(config.refreshToken))
+        return if (config.offline)
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/api/ServiceProvider.kt]: Code style: always wrap blocks with braces

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    override fun cancelStreaming() {
+        streamFuture?.responseReason = CANCELLATION
+        streamFuture?.cancel(true)
+    }
+
+    override fun update(o: Observable?, arg: Any?) {
+        if(o is VADService){
+            println("-----------------------------------------------------------------vAD ----------------------")
```
Comment (ttroy50) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: I know this is just a skeleton implementation but these should probably be log statements

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+class LocalClient(private val config: Config, private val converter: Converter, val vadService : PipelineClient, val acousticService : PipelineClient) : Client, Observer, Observable() {
+
+    init {
+        //PIPELINES connections
+        this.addObserver(vadService)
+        vadService.addObserver(acousticService)
+        acousticService.addObserver(this)
+    }
+
+    lateinit var streamFuture: AudioResponseFuture
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: Should this be private?

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+
+class AssetsLoader(var context: Context ){
+
+//    val vlstmFileName = extractAssets("01DA3PB7N9PWY1147WHSHM0E6W.vlstm", context)
+
+    internal fun extractAssets(fName: String): String {
+        val newLoc = context.getFilesDir().toString().plus(fName)
```
Comment (ttroy50) [inline library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: Is this really needed? I know I had it in the initial test implementation but I didn't investigate if it could be made more efficient that copying the whole file into a new location on disk

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+    override fun sendFeedback(queryId: String, feedback: FeedbackData, token: String): Future<String> {
+        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
+    }
+
+    override fun streamAudio(channel: ReadableByteChannel, queryResponse: QueryResponse): QueryFuture {
+        //new Queries should go to createAudioQuery
+        acousticService.newQuery()
+        vadService.newQuery()
+
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: Unnecessary whitespace

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        streamFuture = AudioResponseFuture()
+        sendLoop(channel, streamFuture)
+
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: Unnecessary whitespace

File: library/src/main/java/com/voysis/sevice/local/loader/ModelLoader.kt [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        val startLoad = System.currentTimeMillis()
+        cLib.load_acoustic_session(modelFile, model)
+        val endLoad = System.currentTimeMillis()
+        val timeLoad = endLoad - startLoad
+        Log.d("ModelLoader", "Acoustic model Loaded, Time to load is $timeLoad")
+
+        return model
+    }
+
+    internal fun loadSpeechModel(modelFile: String): PointerByReference {
```
Comment (ttroy50) [inline library/src/main/java/com/voysis/sevice/local/loader/ModelLoader.kt]: should probably be `loadLanugageModel`

File: library/src/main/java/com/voysis/sevice/local/LocalServiceImpl.kt [hunk 13]
```diff
@@ -0,0 +1,208 @@
+package com.voysis.sevice.local
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/LocalServiceImpl.kt]: I'm going to skip reviewing this file as my understanding is that as dev proceeds it will be removed again such that we have a single `ServiceImpl` implementing `Service` for both cloud and local execution.

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt [hunk 14]
```diff
@@ -0,0 +1,36 @@
+package com.voysis.sevice.local.loader
+
+import android.content.Context
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+
+class AssetsLoader(var context: Context ){
```
Comment (oranoceallaigh) [inline library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: Please add documentation to new classes you introduce. Explain the purpose of the class.

I also think that the `AssetsLoader` should be an interface so that we can use different implementations of the loader: one can obtain assets from the device's local storage but another might choose to download the assets from a remote service.

File: library/src/main/java/com/voysis/sevice/local/loader/ModelLoader.kt [hunk 15]
```diff
@@ -0,0 +1,47 @@
+package com.voysis.sevice.local.loader
+
+import android.util.Log
+import com.sun.jna.Library
+import com.sun.jna.Native
+import com.sun.jna.ptr.PointerByReference
+
+class ModelLoader(var cLib: LoaderLibrary = Native.load("android-test", LoaderLibrary::class.java)) {
```
Comment (ttroy50) [inline library/src/main/java/com/voysis/sevice/local/loader/ModelLoader.kt]: You have `loadVadModel` as part of this class when you create the local service. You can probably leave out vad from from the initial implementation

File: library/src/main/java/com/voysis/sevice/local/modules/AcousticService.kt [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import android.util.Log
+import com.sun.jna.Library
+import com.sun.jna.Native
+import com.sun.jna.Pointer
+import com.sun.jna.ptr.IntByReference
+import com.sun.jna.ptr.PointerByReference
+import java.nio.ByteBuffer
+import java.util.*
+
+class AcousticService(val acousticModel: PointerByReference, val languageModel: PointerByReference, var cLib: AcousticLibrary = Native.load("android-test", AcousticLibrary::class.java)) : PipelineClient() {
```
Comment (ttroy50) [inline library/src/main/java/com/voysis/sevice/local/modules/AcousticService.kt]: Does loading the same library from multiple classes cause any potential issues?

File: library/src/main/java/com/voysis/sevice/local/LocalServiceImpl.kt (hunk 13 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/LocalServiceImpl.kt]: Yes, this is just a temporal solution

File: library/src/main/java/com/voysis/api/ServiceProvider.kt [hunk 17]
```diff
@@ -46,11 +55,47 @@ class ServiceProvider {
...
+                  config: Config,
+                  audioRecorder: AudioRecorder = AudioRecorderImpl(generateAudioWavRecordParams(context, config)),
+                  assetsLoader: AssetsLoader = LocalAssetsLoader(context),
+                  modelLoader: ModelLoader = LocalModelLoader()
+    ): Service {
+        val converter = Converter(getHeaders(context), Gson())
+        return createLocalService(context, config, converter, audioRecorder, assetsLoader, modelLoader)
+    }
+
+    private fun createLocalService(context: Context, config: Config, converter: Converter, audioRecorder: AudioRecorder, assetsLoader: AssetsLoader, modelLoader: ModelLoader): Service {
```
Comment (lDuffy) [inline library/src/main/java/com/voysis/api/ServiceProvider.kt]: I can't see where context is used in this method? is it needed?

File: library/src/main/java/com/voysis/api/ServiceProvider.kt [hunk 18]
```diff
@@ -46,11 +55,47 @@ class ServiceProvider {
...
+        // Load model assets
+        val acousticModelFile = assetsLoader.extractAssets("default-acousticModel.model")
+        val vadAcousticModelFile = assetsLoader.extractAssets("vad-acousticModel.model")
```
Comment (lDuffy) [inline library/src/main/java/com/voysis/api/ServiceProvider.kt]: are these files going to be in the assets folder in the library or the calling application?

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+import com.voysis.sevice.local.modules.PipelineClient
+import com.voysis.sevice.local.modules.VADService
+import java.nio.ByteBuffer
+import java.nio.channels.ReadableByteChannel
+import java.util.Observable
+import java.util.Observer
+import java.util.UUID
+import java.util.concurrent.Future
+
+class LocalClient(private val config: Config, val vadService: PipelineClient, val acousticService: PipelineClient) : Client, Observer, Observable() {
```
Comment (lDuffy) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: should vadService and acousticService also be private?

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+import android.content.Context
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+
+interface AssetsLoader {
+    fun extractAssets(fName: String): String
+}
+
```
Comment (lDuffy) [inline library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: can assetLoader be internal?

File: library/src/main/java/com/voysis/api/ServiceProvider.kt (hunk 18 above)
Comment (oranoceallaigh) [reply library/src/main/java/com/voysis/api/ServiceProvider.kt]: The application. The library will exist as an independent "slim" artifact. If your application wants to use local execution mode, the application build needs to retrieve the necessary assets and include them in the APK such that the library can locate them.

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+class LocalAssetsLoader(var context: Context) : AssetsLoader {
```
Comment (lDuffy) [inline library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: potential memory leak here. We dont want to store a reference to context anywhere in our library if possible. there's an explanation here in this article. https://proandroiddev.com/everything-you-need-to-know-about-memory-leaks-in-android-d7a59faaf46a about half way down

File: library/src/main/java/com/voysis/sevice/local/loader/ModelLoader.kt [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+class LocalModelLoader(var cLib: LoaderLibrary = Native.load("android-test", LoaderLibrary::class.java)) : ModelLoader {
+
+    override fun loadAcousticModel(modelFile: String): PointerByReference {
+        val model = PointerByReference()
+        Log.d("ModelLoader", "Loading Acoustic model...")
+        val startLoad = System.currentTimeMillis()
+        cLib.load_acoustic_session(modelFile, model)
+        val endLoad = System.currentTimeMillis()
```
Comment (jcftang) [inline library/src/main/java/com/voysis/sevice/local/loader/ModelLoader.kt]: this endLoad isn't really necessary, you could just do subtraction below?

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+
+    override fun extractAssets(fName: String): String {
+        val newLoc = context.getFilesDir().toString().plus(fName)
+        val f = File(newLoc)
+        if (!f.exists()) {
+            try {
+                val inputstream = context.getAssets().open(fName)
+                val fileOutputStream = FileOutputStream(f)
+
+                val buf = ByteArray(1024)
```
Comment (jcftang) [inline library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: is 1024 the best value here?

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+                var len: Int = inputstream.read(buf)
+                while (len > 0) {
+                    fileOutputStream.write(buf, 0, len)
+                    len = inputstream.read(buf)
+                }
+                fileOutputStream.close()
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
```
Comment (lDuffy) [inline library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: we should probably throw a specific voysis exception here to indicate that there was a problem with the models the user included in their application.

Comment (jcftang) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: I think this is correct here to be catching the IOException and printing the trace, it is just IO that's happening after all.

Comment (lDuffy) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: I suppose. catching a model exception should probably happen at the point where we try to load the models if that method can throw an error. Wouldn't use `e.printStackTrace` though. it prints to the device logcat without including the calling application. Better to use the android logger. https://stackoverflow.com/a/2073844/2342568

File: library/src/main/java/com/voysis/api/ServiceProvider.kt (hunk 17 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/api/ServiceProvider.kt]: This is the skeleton of what it'll need when c libraries will be integrated

File: library/src/main/java/com/voysis/api/ServiceProvider.kt (hunk 18 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/api/ServiceProvider.kt]: I used some initial assets file names in the skeleton.

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt (hunk 20 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: if it is internal then makeLocal function in ServiceProvider should be internal, and that function is currently being call from MainActivity. 
I think it is not possible currently. We can reconsider refactor how we are calling it if you think that this should be internal

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt [hunk 25]
```diff
@@ -0,0 +1,35 @@
+package com.voysis.sevice.local.modules
+
+import android.util.Log
+import com.sun.jna.Library
+import com.sun.jna.ptr.PointerByReference
+import java.nio.ByteBuffer
+import java.util.Observable
+
+class VADService(val vadAcousticModel: PointerByReference) : PipelineClient() {
```
Comment (jcftang) [inline library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: Could we wrap up VAD and acoustic decoding into one interface on the C++/C side so that this service isn't needed?

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt (hunk 21 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: We have multiple ways to solve this. Could be enough if we store the AssetManager reference?

Comment (lduffy-voysis) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: yeah that sounds good

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt (hunk 23 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: I just copied it. We can changed it to inputstream.available()

Comment (oranoceallaigh) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: Don't use `available`, you'll read the entire model in to memory. Given this is dealing with extracting the assets (only used during init, not during realtime audio streaming) the buffer size is not going to have too much of an impact on overall app performance. 1024 is probably too small but you don't need to spend a lot of time worrying about the size - 4096 or 8192 should be fine, there's probably lower level buffering going on anyway so there isn't a one-to-one correlation to your call to `read` here and actual reads to the underlying storage.

File: library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt (hunk 24 above)
Comment (oranoceallaigh) [reply library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt]: It's not correct to swallow the exception here. You catch an exception if you can do something about it. Here, nothing can be done - the asset extraction failed so the rest of the application cannot run (as it will not be able to load the models.) If you cannot do anything about it, don't catch the exception, let it bubble out so that higher levels in the call stack can deal with it.

File: library/src/main/java/com/voysis/sevice/local/modules/VADService.kt (hunk 25 above)
Comment (aitorrf) [reply library/src/main/java/com/voysis/sevice/local/modules/VADService.kt]: It could be a step that we can consider in a future step.

File: library/src/main/java/com/voysis/sevice/local/LocalClient.kt [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+    private fun checkStreamStopReason(streamFuture: AudioResponseFuture) {
+        if (streamFuture.responseReason != VAD_RECEIVED && streamFuture.responseReason != CANCELLATION) {
+            notifyObservers(END_OF_STREAM)
+            streamFuture.responseReason = END_OF_STREAM
+        }
+    }
+
+    private fun send(buf: ByteBuffer) {
+        setChanged()
+        notifyObservers(buf)
```
Comment (lduffy-voysis) [inline library/src/main/java/com/voysis/sevice/local/LocalClient.kt]: Not sure if this question belongs here but do we have any way of notifying the user of errors that occur from within the models during processing?
